### PR TITLE
3.3.1

### PR DESCRIPTION
**Pull request by richardwestenra**
## Description

- Apply the Black formatter to back-end Python code  (#193)
- Bump websocket-extensions from 0.1.3 to 0.1.4 (#190)
- Autoformat Python code using Black (#188)
- Stop layout calculation from running twice on page-load, and optimise getNodeTextWidth selector performance (#186)
- Fix `%run_viz` line magic to display kedro viz inside Jupyter notebook, previously broken for kedro 0.16.0/0.16.1. Note that the fix needs `kedro>=0.16.2` to take effect. (#184)
- Fix three minor tooltip triangle styling bugs (#182)
- Update readme/contributing docs (#181)
- Update the semver requirement version (#180)
- Seed randomly-generated data with UUIDs, to allow random layouts to be replicated for testing purposes (#178)
- Update pipeline visualisation screenshot in Readme (#177)

## Legal notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -59,35 +59,55 @@ def count_truthy(elements: List[Any]) -> int:
...
+This will serve the app at [localhost:4141](http://localhost:4141/), and watch files in `/src` for changes. It will also update the `/lib` directory, which contains a Babel-compiled copy of the source. This directory is exported to `npm`, and is used when importing as a React component into another application. It is updated automatically when you save in case you need to test/debug it locally (e.g. with `npm link`). You can also update it manually, by running
 
 ```bash
 npm run lib
 ```
 
+#### Data sources
+
+Kedro-Viz uses a unique identifier to determine the data source. You can configure this by appending a query string to the URL, e.g. `http://localhost:4141/?data=random`. Alternatively, you can set it with an environment variable when starting up the dev server:
```
Comment (921kiyo) [inline CONTRIBUTING.md:82]:
```suggestion
Kedro-Viz uses an unique identifier to determine the data source. You can configure this by appending a query string to the URL, e.g. `http://localhost:4141/?data=random`. Alternatively, you can set it with an environment variable when starting up the dev server:
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -59,35 +59,55 @@ def count_truthy(elements: List[Any]) -> int:
...
+There are several different data sources available. By default in production, the app asynchronously loads JSON from the `/api/nodes.json` endpoint, and you can replicate this in development by placing a dataset at `/public/api/nodes.json`. Alternatively, you can use one of the mock unit-testing/demo datasets, or pseudo-random data procedurally-generated on page load, which is often useful for local development.
+
+| Identifier | Data source |
+|------------|-------------|
+| `json` (default) | `/public/api/nodes.json` |
+| `random` | Randomly-generated data |
+| `demo` | `/src/utils/data/demo.mock.js` |
+| `lorem` | `/src/utils/data/lorem-ipsum.mock.js` |
+| `animals` | `/src/utils/data/animals.mock.js` |
+| `layers` | `/src/utils/data/layers.mock.js` |
```
Comment (921kiyo) [inline CONTRIBUTING.md:97]:
```suggestion
| Identifier       | Data source                           |
| ---------------- | ------------------------------------- |
| `json` (default) | `/public/api/nodes.json`              |
| `random`         | Randomly-generated data               |
| `demo`           | `/src/utils/data/demo.mock.js`        |
| `lorem`          | `/src/utils/data/lorem-ipsum.mock.js` |
| `animals`        | `/src/utils/data/animals.mock.js`     |
| `layers`         | `/src/utils/data/layers.mock.js`      |
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -59,35 +59,55 @@ def count_truthy(elements: List[Any]) -> int:
...
-| `docs`           | Changes to the documentation of the plugin                                  |
-| `feature`        | Non-breaking change which adds functionality                                |
-| `fix`            | Non-breaking change which fixes an issue                                    |
-| `tests`          | Changes to project unit (`tests/`) and / or integration (`features/`) tests |
+| Types of changes | Description |
+| ---------------- | ----------- |
+| `docs`    | Changes to the documentation of the plugin |
+| `feature` | Non-breaking change which adds functionality |
+| `fix`     | Non-breaking change which fixes an issue |
+| `tests`   | Changes to project unit (`tests/`) and / or integration (`features/`) tests |
```
Comment (921kiyo) [inline CONTRIBUTING.md:110]:
```suggestion
| Types of changes | Description                                                                 |
| ---------------- | --------------------------------------------------------------------------- |
| `docs`           | Changes to the documentation of the plugin                                  |
| `feature`        | Non-breaking change which adds functionality                                |
| `fix`            | Non-breaking change which fixes an issue                                    |
| `tests`          | Changes to project unit (`tests/`) and / or integration (`features/`) tests |
```

File: README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Kedro-Viz
-| `develop`                                                                                                                                                        | `master`                                                                                                                                                       |
-| ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `develop` | `master`|
+| --------- | ------- |
```
Comment (921kiyo) [inline README.md:3]:
```suggestion
| `develop`                                                                                                                                                        | `master`                                                                                                                                                       |
| ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- |
```

File: README.md [hunk 5]
```diff
@@ -61,15 +61,15 @@ kedro viz
...
-| `--env`, `-e`            | Kedro configuration environment. If not specified, catalog config in `local` will be used.                                                                                             |
+| CLI command | Description |
+| ----------- | ----------- |
+| `--host` | Host that viz will listen to. Defaults to 127.0.0.1. |
+| `--port` | TCP port that viz will listen to. Defaults to 4141. |
+| `--browser/--no-browser` | Whether to open viz interface in the default browser or not. |
+| `--load-file` | Path to load the pipeline JSON file |
+| `--save-file` | Path to save the pipeline JSON file |
+| `--pipeline` | Name of the [modular pipeline](https://kedro.readthedocs.io/en/latest/04_user_guide/06_pipelines.html#modular-pipelines) to visualise. If not set, the default pipeline is visualised. |
+| `--env`, `-e` | Kedro configuration environment. If not specified, catalog config in `local` will be used. |
```
Comment (921kiyo) [inline README.md:72]:
```suggestion
| CLI command              | Description                                                                                                                                                                            |
| ------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `--host`                 | Host that viz will listen to. Defaults to 127.0.0.1.                                                                                                                                   |
| `--port`                 | TCP port that viz will listen to. Defaults to 4141.                                                                                                                                    |
| `--browser/--no-browser` | Whether to open viz interface in the default browser or not.                                                                                                                           |
| `--load-file`            | Path to load the pipeline JSON file                                                                                                                                                    |
| `--save-file`            | Path to save the pipeline JSON file                                                                                                                                                    |
| `--pipeline`             | Name of the [modular pipeline](https://kedro.readthedocs.io/en/latest/04_user_guide/06_pipelines.html#modular-pipelines) to visualise. If not set, the default pipeline is visualised. |
| `--env`, `-e`            | Kedro configuration environment. If not specified, catalog config in `local` will be used.                                                                                             |
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (richardwestenra) [reply CONTRIBUTING.md:97]: This was part of #181 - this is definitely not the appropriate PR to be making these sorts of changes. This is just a release approval PR, we shouldn't be using it to make code/documentation changes other than release notes.

**Comment by richardwestenra** [comment]
Note: Have discussed this with @921kiyo, and we've agreed that we should make a separate PR for these changes, rather than sneak them into a release PR.

**Pull request merged**